### ISSUE TITLE
GOOSE message quality mappings prevent viewing of GOOSE connectivity

### ISSUE THREAD

**Issue by danyill** [openscd/open-scd#297]
Thank you for this project. I am quite inspired and am just beginning to familiarise myself with it.

I was attempting to view GOOSE mappings for an imported scd file (sorry it's not a very minimal reproduction) zipped below:

[IEC61850_GOOSE_2-section.scd.zip](https://github.com/openscd/open-scd/files/7070144/IEC61850_GOOSE_2-section.scd.zip)

When I click on the 'Communication Mapping' icon I receive an error: 

![image](https://user-images.githubusercontent.com/674783/131206537-38e71412-3135-40d6-ac02-6ab35bd83873.png)

![image](https://user-images.githubusercontent.com/674783/131206388-c36c3fc0-82ff-44eb-aad7-3632301228c9.png)

I've traced this to the way in which the keys are generated in `commmap-wizards.ts`. Specifically:

https://github.com/openscd/open-scd/blob/ee932a3139d3c265670a62a02b1a3749e37b06a9/src/wizards/commmap-wizards.ts#L79-L82

If there is no defined controlBlock, then a key with no "identity" or tagName is created.

This generation causes problems shortly after in:

https://github.com/openscd/open-scd/blob/ee932a3139d3c265670a62a02b1a3749e37b06a9/src/wizards/commmap-wizards.ts#L91-L102

which errors out on the `querySelector` because the selector doesn't exist. The cause is that

* `cbTag` does not exist
* `cbId` does not exist

and therefore there is no valid DOM match.

I attribute this to the mapping of overall message quality, as in the following:

```xml
<ExtRef iedName="SSS2022P3" intAddr="VB004|msgQ=true" serviceType="GOOSE" srcLDInst="CFG" srcLNClass="LLN0" srcCBName="TripArcDetBusK" />
<ExtRef iedName="SSS2022P3" ldInst="ANN" prefix="SVT" lnClass="GGIO" lnInst="4" doName="Ind02" daName="stVal" intAddr="VB084|q=0x1FFF" serviceType="GOOSE" srcLDInst="CFG" srcLNClass="LLN0" srcCBName="TripArcDetBusK" />
<ExtRef iedName="SSS2022P3" ldInst="ANN" prefix="SVT" lnClass="GGIO" lnInst="4" doName="Ind03" daName="stVal" intAddr="VB104|q=0x1FFF" serviceType="GOOSE" srcLDInst="CFG" srcLNClass="LLN0" srcCBName="TripArcDetBusK" />
<ExtRef iedName="SSS2022P3" ldInst="ANN" prefix="SVT" lnClass="GGIO" lnInst="4" doName="Ind21" daName="stVal" intAddr="VB024|q=0x1FFF" serviceType="GOOSE" srcLDInst="CFG" srcLNClass="LLN0" srcCBName="TripArcDetBusK" />
<ExtRef iedName="SSS2022P3" ldInst="ANN" prefix="SVT" lnClass="GGIO" lnInst="4" doName="Ind22" daName="stVal" intAddr="VB044|q=0x1FFF" serviceType="GOOSE" srcLDInst="CFG" srcLNClass="LLN0" srcCBName="TripArcDetBusK" />
<ExtRef iedName="SSS2022P3" ldInst="ANN" prefix="SV" lnClass="GGIO" lnInst="3" doName="Ind23" daName="stVal" intAddr="VB064|q=0x1FFF" serviceType="GOOSE" srcLDInst="CFG" srcLNClass="LLN0" srcCBName="TripArcDetBusK" />
``` 

The first item here is the mapping of the overall message quality and because there are no control block parameters the error is occurring.

Depending on how overall message quality should be displayed in the UI, a reasonable option may be simply to not show these for now. I found I could view the mappings if I just "skipped" the `querySelector` if `cbTag` and `cbId` were falsy.

**Comment by JakobVogelsang** [comment]
Hi and thank you for this detail issue you have reported. We are very happy about everyone using this software. 

Regarding the bug that you have reported. I have tested it already and should be capable to fix it within the next couple of days. I agree with you, the simplest solution would be to simply not show `ExtRef` elements that do not satisfy some retirements.

Just for curiosity. Do you need to manipulate the first `ExtRef` element above, and what is it used for?

**Comment by danyill** [comment]
Thanks!

Regarding what:

```js
<ExtRef iedName="SSS2022P3" intAddr="VB004|msgQ=true" serviceType="GOOSE" srcLDInst="CFG" srcLNClass="LLN0" srcCBName="TripArcDetBusK" />
```
is used for:

This output is produced by the vendor specific tool (in this case, SEL Architect) and is used to reproduce the internal mapping of the "overall GOOSE message quality" onto an internal address (in this case the internal address is called `VB004`). 

This "overall GOOSE message quality" means (I think) if there is `DetectedPacketLoss` (see Fig 11 of IEC 61850-8-1 Ed 2.1) and perhaps also the `Non-existent` state.

![image](https://user-images.githubusercontent.com/674783/131313117-04ad76a3-5b33-4358-a199-1c2392df8546.png)

While I believe this SCD file is valid, this mapping behaviour may be manufacturer-specific, the standard says: "The handling and processing of received GOOSE messages, by the subscriber, is a local issue." (IEC 61850-8-1 Ed 2.1b, 18.1.2.5.1).

Here's a screenshot of how the item above looks in the vendor tool:

![image](https://user-images.githubusercontent.com/674783/131313682-44d56f40-3e4e-48ac-876f-7295341dc069.png)

**Comment by ca-d** [comment]
Thank you for this extraordinarily clear and detailed model of a bug report. This is truly the epitome of what one could wish for from a passionately engaged open source user 🤘😎

**Comment by ca-d** [comment]
@JakobVogelsang resolved this for now in #298 by simply ignoring ExtRef elements without a valid source. There might be some more flexible way of handling this case in the future.
